### PR TITLE
http: select stable stream ID source based on route table

### PR DESCRIPTION
**Pull request by mattklein123**
We've gone through a few different iterations of how we select the stable
random ID uses for request routing. We used to use a random number for all
requests, but this yields a performance hit when there is no need for a
true random number. Then we switched to an incrementing ID which gives
approximate randomness at high RPS, but is confusing for new users and in
certain cases does not really provide the expected behavior.

In this commit we implement both cases. If the route table does not use
runtime we use the fast approach. If it uses runtime, we use a real random
number which will provide behavior that most people would expect.

### CODE REVIEW

**Comment by mattklein123** [comment]
@lyft/network-team @rshriram

**Comment by louiscryan** [comment]
Matt,

I don't know if Enrico reached out to describe the technique we frequently
use at Google to address the perf issues with PRNG for UID like issues?

- Louis

**Comment by mattklein123** [comment]
@louiscryan no, he didn't. Any help/guidance appreciated on this.

**Comment by louiscryan** [comment]
tldr;

we don't allocate one id at a time, we allocate a very large block of them
into a contiguous memory block all at once and then we slice out of that
over time. In a low thread count system its thread-local, in high thread
count systems the slicing is done under lock or by using atomics. All
fairly typical cost amortization stuff.

Using a faster PRNG to allocate the block also helps substantially. E.g.

http://www.pcg-random.org/download.html

**Comment by mattklein123** [comment]
@louiscryan interesting. Thanks, makes sense. I will open a different issue to track improving how we do PRNG.
